---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# STRBaseclient

<!-- badges: start -->
<!-- badges: end -->

STRBaseclient is an R-package for accessing [NIST](https://www.nist.gov/)'s [STRBase](https://strbase.nist.gov/) API. STRBase hosts information about Short Tandem Repeats (STRs) relevant to the human identity community. Resources include STR general information, observations of variant alleles by the forensic community, STR typing kits, and population data among other topics. See [Introduction to STRBase](https://strbase.nist.gov/Others/Intro) for more information.

Note that the author of STRBaseclient is not affiliated with NIST.

## Installation

You can install the development version of STRBaseclient from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("mkruijver/STRBaseclient")
```

## Usage

First we pull a list of loci, similar to the left-hand side of the landing page of STRBase:
```{r}
library(STRBaseclient)

locus_category_table <- get_locus_category_table()
head(locus_category_table)
```

Let's pull the table of variant alleles observed at the first locus (CSF1PO):

```{r}
locus_variant_alleles <- get_locus_variant_alleles_table("CSF1PO")
head(locus_variant_alleles)
```

A similar function retrieves the tri-alleles table:
```{r}
tri_alleles <- get_locus_tri_alleles_table("CSF1PO")
head(tri_alleles)
```

The *get_locus_chromosome_location_and_allele_reference_table* function can be used to look up the location of a locus with respect to the reference genomes:

```{r}
get_locus_chromosome_location_and_allele_reference_table("CSF1PO")
```

## Advanced usage

The *get_...* functions return DataFrames with roughly the same information as the STRBase website. The API often returns a richer dataset which can be inspected when calling a function with the *return_raw_data* argument set to *TRUE*. For example:

```{r}
locus_locations <- get_locus_chromosome_location_and_allele_reference_table("CSF1PO", return_raw_data = TRUE)
str(locus_locations)
```

It is also possible to query an arbitrary API endpoint using the *get_endpoint* function:

```{r}
locus <- "SE33"
  
endpoint <- paste0("LociByName/Loci/", locus)
response <- get_endpoint(endpoint)

response
```

To retrieve the raw JSON response, use the *get_endpoint_JSON* function:


```{r}
response_JSON <- get_endpoint_JSON(endpoint)

response_JSON
```


